ego/cli/bundle.go

```diff
@@ -0,0 +1,240 @@
+func (c *Cli) buildImage(enclaveFilename string) (string, error) {
+	// Copy the runtime image to the tarPreparedContent
+	egoHostPath := c.getEgoHostPath()
+	egoEnclavePath := filepath.Join(c.egoPath, "share", "ego-enclave")
+
+	// Create the tar file
+	tempFile, err := c.fs.TempFile("", "ego-runtime")
+	if err != nil {
+		return "", err
+	}
+	defer tempFile.Close()
+
+	// Setup tar writer with gzip compression
+	compressedWriter := gzip.NewWriter(tempFile)
+	defer compressedWriter.Close()
+	tarWriter := tar.NewWriter(compressedWriter)
+	defer tarWriter.Close()
+
+	// Add the runtime image to the tar file
+	egoHostTargetPath := filepath.Join("bin", "ego-host")
+	if err := c.addToArchive(tarWriter, egoHostPath, egoHostTargetPath); err != nil {
+		return "", err
+	}
+	egoEnclaveTargetPath := filepath.Join("share", "ego-enclave")
+	if err := c.addToArchive(tarWriter, egoEnclavePath, egoEnclaveTargetPath); err != nil {
+		return "", err
+	}
+
+	// Add the enclave to the tar file
+	if err := c.addToArchive(tarWriter, enclaveFilename, "enclave"); err != nil {
+		return "", err
+	}
+
+	return tempFile.Name(), err
+}
+
+// addToArchive adds a file from the filesystem to the tar archive.
+func (c *Cli) addToArchive(tw *tar.Writer, sourceFilename string, targetFilename string) error {
+	file, err := c.fs.Open(sourceFilename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	sourceFileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	header, err := tar.FileInfoHeader(sourceFileInfo, sourceFileInfo.Name())
+	if err != nil {
+		return err
+	}
+
+	header.Name = targetFilename
+
+	if err := tw.WriteHeader(header); err != nil {
+		return err
+	}
+
+	n, err := io.Copy(tw, file)
+	if err != nil {
+		return err
+	}
+
+	if n != sourceFileInfo.Size() {
+		return fmt.Errorf("file size mismatch: %d written instead of expected %d bytes", n, sourceFileInfo.Size())
+	}
+
+	return nil
+}
+
+// prepareBundle copies the "empty" bundle executable to the target location and names it after the enclave to be bundled.
+func (c *Cli) prepareBundle(filename string) (string, error) {
+	// Create new file which eventually ends up as the packed binary
+	filenameBase := filepath.Base(filename)
+	filenameBaseNoExt := strings.Split(filenameBase, ".")[0]
+	binaryToPack, err := c.fs.Create(filenameBaseNoExt + "-bundle")
+	if err != nil {
+		return "", err
+	}
+	defer binaryToPack.Close()
+
+	// Appereantly, go build creates binaries with 775... So let's just replicate this.
+	if err := c.fs.Chmod(filenameBase+"-bundle", 0o775); err != nil {
+		return binaryToPack.Name(), err
+	}
+
+	// Copy the base bundle loader binary into the new file
+	loaderBinaryPath := filepath.Join(c.egoPath, "bin", "ego-bundle")
+	loaderBinary, err := c.fs.Open(loaderBinaryPath)
+	if err != nil {
+		return binaryToPack.Name(), err
+	}
+	defer loaderBinary.Close()
+
+	loaderBinaryInfo, err := loaderBinary.Stat()
+	if err != nil {
+		return binaryToPack.Name(), err
+	}
```

Comment:
nit: I'd do these readonly operations before the create operation to not (even temporarily) change system state on err. (But not really a big deal here,)